instrumentation/github.com/gocql/gocql/observer.go

```diff
@@ -0,0 +1,263 @@
+
+		queryLabels := append(
+			defaultMetricLabels(keyspace, host),
+			cassStatement(observedQuery.Statement),
+		)
+		iQueryCount.Add(
+			ctx,
+			1,
+			queryLabels...,
+		)
+		iQueryRows.Record(
+			ctx,
+			int64(observedQuery.Rows),
+			defaultMetricLabels(keyspace, host)...,
+		)
+		iLatency.Record(
+			ctx,
+			nanoToMilliseconds(observedQuery.Metrics.TotalLatency),
+			defaultMetricLabels(keyspace, host)...,
+		)
+	}
+
+	if o.observer != nil {
+		o.observer.ObserveQuery(ctx, observedQuery)
+	}
+}
+
+// ObserveBatch is called once per batch query, and provides instrumentation for it.
+func (o *OtelBatchObserver) ObserveBatch(ctx context.Context, observedBatch gocql.ObservedBatch) {
+	if o.cfg.InstrumentBatch {
+		host := observedBatch.Host
+		keyspace := observedBatch.Keyspace
+		attributes := append(
+			defaultAttributes(host),
+			cassBatchQueries(len(observedBatch.Statements)),
+			cassKeyspace(keyspace),
+		)
+
+		ctx, span := o.cfg.Tracer.Start(
+			ctx,
+			cassBatchQueryName,
+			trace.WithStartTime(observedBatch.Start),
+			trace.WithAttributes(attributes...),
+		)
+
+		if observedBatch.Err != nil {
+			span.SetAttributes(cassErrMsg(observedBatch.Err.Error()))
+			recordError(ctx, iBatchErrors, keyspace, host)
+		}
+
+		span.End(trace.WithEndTime(observedBatch.End))
+
+		iBatchCount.Add(
+			ctx,
+			1,
+			defaultMetricLabels(observedBatch.Keyspace, observedBatch.Host)...,
+		)
+		iLatency.Record(
+			ctx,
+			nanoToMilliseconds(observedBatch.Metrics.TotalLatency),
+			defaultMetricLabels(keyspace, host)...,
+		)
+	}
+
+	if o.observer != nil {
+		o.observer.ObserveBatch(ctx, observedBatch)
+	}
+}
+
+// ObserveConnect is called once per connection attempt, and provides instrumentation for it.
+func (o *OtelConnectObserver) ObserveConnect(observedConnect gocql.ObservedConnect) {
+	if o.cfg.InstrumentConnect {
+		host := observedConnect.Host
+		hostname := getHost(host.HostnameAndPort())
+		attributes := defaultAttributes(observedConnect.Host)
+
+		_, span := o.cfg.Tracer.Start(
+			o.ctx,
+			cassConnectName,
+			trace.WithStartTime(observedConnect.Start),
+			trace.WithAttributes(attributes...),
+		)
+
+		if observedConnect.Err != nil {
+			span.SetAttributes(cassErrMsg(observedConnect.Err.Error()))
+			iConnectErrors.Add(
+				o.ctx,
+				1,
+				cassHost(hostname),
+				cassHostID(host.HostID()),
+			)
+		}
+
+		span.End(trace.WithEndTime(observedConnect.End))
+
+		iConnectionCount.Add(
+			o.ctx,
+			1,
+			cassHost(hostname),
+			cassHostID(host.HostID()),
+		)
+	}
+
+	if o.observer != nil {
+		o.observer.ObserveConnect(observedConnect)
+	}
+}
+
+// ------------------------------------------ Private Functions
+
+// getHost returns the hostname as a string.
+// gocql.HostInfo.HostnameAndPort() returns a string
+// formatted like host:port. This function returns the host.
+func getHost(hostPort string) string {
+	idx := strings.Index(hostPort, ":")
+	host := hostPort[0:idx]
+	return host
+}
```

Comment:
Could [`net.SplitHostPort`](https://golang.org/pkg/net/#SplitHostPort) replace this or be used instead of `string.Index`?

Reply:
Yep!
With my experimentation using `localhost`, `gocql.HostnameAndPort()` appears to return a resolved IP and port number (`127.0.0.1:9042`). So I thought that I'd do something along the lines of this:
```go
func hostOrIP(hostnameAndPort string) kv.KeyValue {
	hostname, _, err := net.SplitHostPort(hostnameAndPort)
	if err != nil {
		log.Printf("failed to parse hostname from port, %v", err)
	}
	if parse := net.ParseIP(hostname); parse != nil {
		return cassPeerIP(parse.String())
	}
	return cassPeerName(hostname)
}
```